misc/memory-connection-limits/src/lib.rs

```diff
@@ -0,0 +1,231 @@
+// DEALINGS IN THE SOFTWARE.
+
+use libp2p_core::{Endpoint, Multiaddr};
+use libp2p_identity::PeerId;
+use libp2p_swarm::{
+    dummy, ConnectionDenied, ConnectionId, FromSwarm, NetworkBehaviour, PollParameters, THandler,
+    THandlerInEvent, THandlerOutEvent, ToSwarm,
+};
+use void::Void;
+
+use std::{
+    fmt,
+    task::{Context, Poll},
+};
+
+/// A [`NetworkBehaviour`] that enforces a set of memory usage based limits.
+///
+/// For these limits to take effect, this needs to be composed into the behaviour tree of your application.
+///
+/// If a connection is denied due to a limit, either a [`SwarmEvent::IncomingConnectionError`](libp2p_swarm::SwarmEvent::IncomingConnectionError)
+/// or [`SwarmEvent::OutgoingConnectionError`](libp2p_swarm::SwarmEvent::OutgoingConnectionError) will be emitted.
+/// The [`ListenError::Denied`](libp2p_swarm::ListenError::Denied) and respectively the [`DialError::Denied`](libp2p_swarm::DialError::Denied) variant
+/// contain a [`ConnectionDenied`](libp2p_swarm::ConnectionDenied) type that can be downcast to [`MemoryUsageLimitExceeded`] error if (and only if) **this**
+/// behaviour denied the connection.
+///
+/// If you employ multiple [`NetworkBehaviour`]s that manage connections, it may also be a different error.
+///
+/// # Example
+///
+/// ```rust
+/// # use libp2p_identify as identify;
+/// # use libp2p_swarm_derive::NetworkBehaviour;
+/// # use libp2p_memory_connection_limits as memory_connection_limits;
+///
+/// #[derive(NetworkBehaviour)]
+/// # #[behaviour(prelude = "libp2p_swarm::derive_prelude")]
+/// struct MyBehaviour {
+///   identify: identify::Behaviour,
+///   limits: memory_connection_limits::Behaviour
+/// }
+/// ```
+pub struct Behaviour {
+    max_process_memory_usage_bytes: Option<usize>,
+    max_process_memory_usage_percentage: Option<f64>,
+    system_physical_memory_bytes: usize,
+}
+
+impl Behaviour {
+    /// Sets the process memory usage threshold in absolute bytes.
+    ///
+    /// New inbound and outbound connections will be denied when the threshold is reached.
+    pub fn with_max_bytes(bytes: usize) -> Self {
+        let mut b = Self::new();
+        b.update_max_bytes(bytes);
+        b
+    }
+
+    /// Sets the process memory usage threshold in the percentage of the total physical memory,
+    /// all pending connections will be dropped when the threshold is exeeded
+    pub fn with_max_percentage(percentage: f64) -> Self {
+        let mut b = Self::new();
+        b.update_max_percentage(percentage);
+        b
+    }
+
+    /// Updates the process memory usage threshold in bytes,
+    pub fn update_max_bytes(&mut self, bytes: usize) {
+        self.max_process_memory_usage_bytes = Some(bytes);
+    }
+
+    /// Updates the process memory usage threshold in the percentage of the total physical memory,
+    pub fn update_max_percentage(&mut self, percentage: f64) {
+        self.max_process_memory_usage_percentage = Some(percentage);
+    }
```

Comment:
Do you feel strongly about keeping these? I think they are redundant because a user can always do the following:

1. To have both limits active:

 ```rust
use libp2p_swarm_derive::NetworkBehaviour;
use libp2p_memory_connection_limits as memory_connection_limits;

#[derive(NetworkBehaviour)]
struct MyBehaviour {
   absolute_limits: memory_connection_limits::Behaviour,
   relative_limits: memory_connection_limits::Behaviour
}

fn main() {
	let _ = MyBehaviour {
		absolute_limits: memory_connection_limits::Behaviour::with_max_bytes(100000),
		relative_limits: memory_connection_limits::Behaviour::with_max_percentage(0.80),
	};
}
```

2. To update certain limits:

```rust
fn main() {
	let mut swarm = todo!();

	swarm.behaviour_mut().limits = memory_connection_limits::Behaviour::with_max_bytes(100000);
}
```

They need a `&mut`  reference to call these update functions so they can also just assign a new value to the field.

What do you think?

Reply:
TBH these methods are only useful for the unit tests as of now.